[24.2] Guide users to collection builders

ahmedhamidawan opened the pull request:
### Major action items (_UPDATE 11/19 now complete for this PR_):
- [x] _By default, instead of the modals treating the items as selections from the current history, automatically filter items valid for the list (e.g.: for a list with csv elements, filter out csvs from the history in this list)_ 
  ✅ **Done: If there are required extension(s), we filter out the items with the required extension (but we also include any items that have extension that `isSubTypeOfAny`; meaning it can be implicitly converted to become a valid input).**
- [x] In case nothing can be auto paried for `list:paired`, do not attempt to auto pair by default and simply show all items.
- [ ] ~In case the current history is empty and to make it clearer in general, allow history to be switched from within the modal?~ _Seems like a bad idea, history can just be switched from the Switch history modal; and in the case of a history-less (singular history) UI, this won't be a concern either_
- [x] Allow files to be uploaded (and dropped) directly to either the form field or within the list builder once it is opened.
 ✅ **Done: The collection creator (for all types: `list`, `paired`, `list:paired`) allows users to upload datasets (even binding those to the required extension(s)) directly within itself, and even adds them to the resulting list (only in the case of `list`, for the rest they are shown in the updated available datasets to add list).**

~~_Note: `paired` collection is still a WIP_~~ Done ✅ 

Fixes https://github.com/galaxyproject/galaxy/issues/18704

https://github.com/user-attachments/assets/f0ac9b66-589e-44e5-b2c3-df7f7da04504

### Future improvements incoming:
Maybe we can merge this for the 24.2 release and then I can build upon these changes with the following enhancements:

- Biggest update will be when we add a **"Create your own collection type"** interface, which allows users to create a collection of any shape; currently only `list`, `list:paired` and `paired` are supported/improved by these changes. It would be cool if users can create a shape like `list:list:paired` on the fly. _Please let me know if this is a good idea and/or what are other possibilities that can empower users further here._
- This took A LOT of refactoring of collection creator components and we have quite a bit of duplication in each of these; meaning we would benefit from these files being refactored into reusable components (maybe even a singular component).
- _There are several little_ `TODO`s _here and there right in the code changes in this PR which would improve UX and the code itself_

<details>
<summary>
<h3><i>Outdated screencast</i></h3>
</summary>

https://github.com/user-attachments/assets/8882915e-d0ed-4d5c-a1e6-45eb30b9b3e3

</details>

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
add button for creating a list from run form field

This is an initial/draft implementation. Some of the next steps are:
- By default, instead of the modals treating the items as selections from the current history, automatically filter items valid for the list (e.g.: for a list with csv elements, filter out csvs from the history in this list).
- In case nothing can be auto paried for `list:paired`, do not attempt to auto pair by default and simply show all items.
- In case the current history is empty and to make it clearer in general, allow history to be switched from within the modal?
- Allow files to be uploaded (and dropped) directly to either the form field or within the list builder once it is opened.

One thing I have not planned for yet is the rule builder. I can see that for `list` and `list:paired`, we get that from the `props.collectionTypes` in `FormData`. But when would we use the rule builder instead?

Fixes https://github.com/galaxyproject/galaxy/issues/18704

ahmedhamidawan committed:
fully implement `list` collection creator in `FormData`

This allows a collection type `list` to be created via the collection creater from the workflow/tool form directly.
It tracks the current history changes via the new `useHistoryItemsForType` composable.
It utilises the `FormSelectMany` component to easily move items between selected and unselected for list columns.
The items in the list creator can be filtered for extension, parent datatype or all items in the history, based on whether the form field required a certain extension(s) as input for the list.

ahmedhamidawan committed:
add a `maintain-selection-order` prop to `FormSelectMany`

This keeps the order in which the user adds items to the selection evident in the selected column.

mvdbeek commented on client/src/components/Datatypes/model.ts:
```diff
@@ -39,4 +39,14 @@ export class DatatypesMapperModel {
     }
+
+    /** For classes like `galaxy.datatypes.{parent}.{extension}`, get the extension's parent */
+    getParentDatatype(extension: string) {
```
That logic is redundant and also doesn't reflect how datatypes match. Is there a problem with using `isSubTypeOfAny` ?

mvdbeek commented on client/src/components/Collections/ListCollectionCreator.vue:
```diff
@@ -131,6 +185,18 @@ function _isElementInvalid(element: HistoryItemSummary) {
+            // is the element's extension not a subtype of any of the required extensions?
+            (datatypeToggle.value === "ext" &&
+                !datatypesMapper.value?.isSubTypeOfAny(element.extension, props.extensions!)) ||
+            // else, does the element's extension have the same parent class as any of the required extensions?
```
I don't think this is correct, `isSubTypeOfAny` should take this into account already, otherwise datatype matching in the workflow editor or drag and drop extension validation would be broken. Do you have an example where isSubTypeOfAny returns false and isSubClassOfAny is true ?

ahmedhamidawan replied:
Initially, I had enforced the usage of `isSubTypeOfAny` and hence, the collection builder would only be filtered for extensions explicitly defined (`isSubTypeOfAny === true`).

But then I realized that, for e.g.: if the required extension for the list is `csv`, if you manually create a list from datasets that are not `csv` but are `tabular` instead, that _will_ be a valid input for the csv list **via implicit conversion**.

Therefore, I looked for a way to be more lenient in the list creation as even if the required extension is `csv` and you have no explicit `csv`s, you can still create a list that could be valid via implicit conversion. I felt that the `datatypesMapper` could be the solution for this?

| Here is an example of this: |
| ---- |
| <video src="https://github.com/user-attachments/assets/ff5a4ae3-31f5-47c4-9e88-c65d20641364" /> |
| Here, you can see that for the `Extension` filter, we only show the csvs in the history (`isSubTypeOfAny`), for `Datatype` we show other `tabular` types that you could still create a collection from (`isSubClassOfAny`), and `All` allows for any extension. Then I show you how if you manually create a list from `tabular`s instead, it can still be a valid input... |

> datatype matching in the workflow editor or drag and drop extension validation would be broken

You make a good point though, because once that collection (from extensions other than the required one) is created, I enforce it as input for the field, and that does not seem right because the implicit conversion didn't even happen at that point (I just let the `tabular` list be an input for the field).


If this filtering is problematic, I can do one of these two things (unless there is any other recommendation):
1. Remove the filter buttons, and enforce only `isSubTypeOfAny`; e.g.: only show csvs for csv requirement, and let the user know: "Hey, you might have other datasets available here that you can convert to csv yourself if you wish to add them to the list..."
2. Remove the filter buttons, and extension filtering entirely: allowing users to make a list from whatever, and then letting `FormData` do the validation itself

_Both of these options would mean removing the logic of `isSubClassOfAny` of course._

### Update:
I have removed this optional choice between the level of filtering and enforced only allowed extensions (and others that are `isSubTypeOfAny`.

ahmedhamidawan committed:
fix reactivity of `ClickToEdit` and add some styling

ahmedhamidawan committed:
`ListCollectionCreator`: add more types

ahmedhamidawan committed:
modernize/refactor `PairedListCollectionCreator` for input forms

- Converted the file(s) to composition API and typescript
- Improved styling of the modal and its components

Still need to add `extensions` handling for cases where a certain extension is required for the collection.

ahmedhamidawan committed:
remove the extensions toggle, only include `isSubTypeOfAny` items

The `isSubClassOfAny` was incorrect logic for implicit conversions. `isSubTypeOfAny` gives us what we want as far as filtering items that would be valid for implicit conversions goes.
Also, we concluded that the `All` option was also not acceptable as only valid extensions must be enforced in the collection creator.

ahmedhamidawan committed:
improve styling of collection create button in `FormData`

Place it right next to the buttons for choosing between batch or singular collection

ahmedhamidawan committed:
change create new collection `ButtonSpinner` variant

ahmedhamidawan committed:
change create new collection `ButtonSpinner` title; fix icon imports

ahmedhamidawan committed:
add history name to modal header

ahmedhamidawan committed:
restrict the `list:paired` creator to required extensions

ahmedhamidawan committed:
make the "pairing" section in `list:paired` builder expand/collapseable

ElectronicBlueberry commented on client/src/components/Form/Elements/FormSelectMany/worker/selectManyMain.ts:
```diff
@@ -43,6 +43,16 @@ export function main(options: UnwrapNestedRefs<UseSelectManyOptions>): UnwrapNes
+        const selectedValuesArray = Array.from(selectedValues);
+        selectedOptionsFiltered.sort((a, b) => {
+            const aIndex = selectedValuesArray.findIndex((v) => v === stringifyObject(a.value));
+            const bIndex = selectedValuesArray.findIndex((v) => v === stringifyObject(b.value));
```
Slight optimization.
`stringifyObject` does not need to be re-run for every selected value.

```suggestion
            const aAsString = stringifyObject(a.value);
            const bAsString = stringifyObject(b.value);

            const aIndex = selectedValuesArray.findIndex((v) => v === aAsString);
            const bIndex = selectedValuesArray.findIndex((v) => v === bAsString);
```

ahmedhamidawan committed:
add `CollectionCreatorModal` that replaces `buildCollectionModal`

- `buildCollectionModal.ts` still exists but just to create the rules collection modal, all other modals are replaced with a parent `CollectionCreatorModal`
- also added a `collectionBuilderItemsStore` that is used to store the datasets fetched for the builder; only when the builder is opened: which is when we start reacting to any changes in the `historyId`, `history.update_time` and any filter on a history selection.

ahmedhamidawan committed:
slight optimization

`stringifyObject` does not need to be re-run for every selected value.

Co-authored-by: Laila Los <44241786+ElectronicBlueberry@users.noreply.github.com>

ahmedhamidawan committed:
(incomplete/WIP) add uploader to collection creator

ahmedhamidawan committed:
`paired` collection items are now also reversed like history

ahmedhamidawan committed:
add uploaded files to collection directly

In the case of `list`, they are now added directly to the final list.
In the case of `list:paired` and `paired`, they are added to the `workingElements`; the user can then manually add them.
This is all still done within the `CollectionCreator`.

Next thing to look at is to try to only allow/wait for those uploaded files to be attempted as collection additions if they are of `ok` state.
One way to do that is to maybe use a list to keep track of uploaded items in the creator, and if those uploaded items disappear from `workingElements` (meaning they are invalid) we have explicit Toasts mentioning this.

dannon committed:
Fix jest collection modal test to be more flexible, allowing title content variability

ahmedhamidawan committed:
remove legacy JQuery collection creator modal files

ahmedhamidawan committed:
fix `PairedListCollectionCreator` jest

jmchilton committed:
This really stablizes these tests - they sort of runaway without this.

ElectronicBlueberry committed:
fix pesky warning

ahmedhamidawan committed:
fix selectors for the changes to the collection builder

fix selector for name collection in collection builder

fix hide original items selector in collection creator

fix clear filters selector in collection creator

ahmedhamidawan committed:
do not show create another collection link if from selection

ahmedhamidawan committed:
fix selection for collection creator in libraries

ahmedhamidawan committed:
fix reset function when creating collection from a selection

ahmedhamidawan committed:
fix `selection` object for rule collection builder

ahmedhamidawan committed:
ensure library import to collection modal closes

This fixes seleniums failing in `lib/galaxy_test/selenium/test_library_to_collections.py::TestLibraryToCollections` where the modal was in the way.
